### ISSUE TITLE
mixed precision for non-Keras TensorFlow scripts

### ISSUE THREAD

**Issue by sumanthratna** [tensorflow/tensorflow#38724]
**System information**
- TensorFlow version (you are using): 2.1.0
- Are you willing to contribute it (Yes/No): No (it would take too long for me to learn the low-level functionality of mixed-precision)

**Describe the feature and the current behavior/state.**
Currently, https://www.tensorflow.org/guide/keras/mixed_precision only works with Keras. 

**Will this change the current API? How?**
The mixed_precision module will either be relocated so that it is not exclusively for Keras, or a new submodule will be created.

**Who will benefit with this feature?**
Users who want to use mixed-precision but aren't using Keras.

**Any Other info.**
I'm not sure if this is even possible.

**Comment by lgeiger** [comment]
I think you should be able to use: https://www.tensorflow.org/api_docs/python/tf/train/experimental/enable_mixed_precision_graph_rewrite

**Comment by sumanthratna** [comment]
`tf.train.experimental.enable_mixed_precision_graph_rewrite` returns an instance of `tf.keras.optimizers.Optimizer`, so I'm not sure if this is what I want. If it *does* work without Keras, this should definitely be added to the documentation.

**Comment by reedwm** [comment]
We have considered having a non-Keras mixed precision API but it probably won't happen. Having an easy-to-use non-Keras API would require automatically inserting casts before each op, which is fairly complicated and would negate any user casts. Still, the final decision has not been made and there will be a mixed precision RFC. Once the RFC is available, you and others can voice you concerns with it, including the fact it is Keras only.

For `enable_mixed_precision_graph_rewrite`, you are required to use an optimizer, and in TF 2, only Keras optimizers are available. Other than Keras optimizers, you do not need Keras. Technically, you can also pass a dummy optimizer and not use it, as long as you are careful to implement loss scaling. But we plan on deprecating `enable_mixed_precision_graph_rewrite` soon, so it's not a good option in general.

It is possible to use mixed precision outside Keras by manually inserting casts in the appropriate places. Variables must be created in float32 then casted to float16, and the model inputs must be casted to float16 as well. Loss scaling must also be done.

I'll keep this issue open until the RFC has been up for review for a few weeks and approved.

**Comment by sumanthratna** [comment]
Thanks for your detailed reply, @reedwm! Can you explain what happens if I do the following in a script that doesn't use Keras? 
```python
tf.config.optimizer.set_experimental_options({
    'auto_mixed_precision': True
})
```
Will it simply not do anything?

**Comment by reedwm** [comment]
That `tf.config.optimizer.set_experimental_options` code is equivalent to calling `enable_mixed_precision_graph_rewrite`, except it doesn't wrap an optimizer with a [LossScaleOptimizer](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/LossScaleOptimizer). It will affect non-Keras code since it rewrites the graph to use mixed precision and doesn't care about whether Keras is used or not.

However, that code is dangerous since it doesn't do any loss scaling. You will either have to do loss scaling by hand, or, if you use a Keras optimizer, wrap the optimizer with a LossScaleOptimizer. In the latter case, you should use `enable_mixed_precision_graph_rewrite` instead of `tf.config.optimizer.set_experimental_options` to automatically have the optimizer wrapped with a LossScaleOptimizer.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by sumanthratna** [comment]
@reedwm Seems like this issue might still be relevant, correct? (given that no non-Keras API exists for mixed precision [and there was no RFC on this topic, to my knowledge])

**Comment by Venkat6871** [comment]
Hi @sumanthratna ,
Thank you so much for filing this feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
